registry/distribution/repository.go

```diff
@@ -0,0 +1,674 @@
+	}
+	return ocispec.Descriptor{
+		MediaType: mediaType,
+		Digest:    refDigest,
+		Size:      size,
+	}, nil
+}
+
+// manifestStore accesses the manifest part of the repository.
+type manifestStore struct {
+	repo *Repository
+}
+
+// Fetch fetches the content identified by the descriptor.
+func (s *manifestStore) Fetch(ctx context.Context, target ocispec.Descriptor) (rc io.ReadCloser, err error) {
+	url := fmt.Sprintf("%s/manifests/%s", s.repo.endpoint(), target.Digest)
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Accept", target.MediaType)
+
+	resp, err := s.repo.Client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			resp.Body.Close()
+		}
+	}()
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+		// no-op
+	case http.StatusNotFound:
+		return nil, fmt.Errorf("%s: %w", target.Digest, errdef.ErrNotFound)
+	default:
+		return nil, parseErrorResponse(resp)
+	}
+	mediaType, _, err := mime.ParseMediaType(resp.Header.Get("Content-Type"))
+	if err != nil {
+		return nil, fmt.Errorf("%s %q: invalid response Content-Type: %w", resp.Request.Method, resp.Request.URL, err)
+	}
+	if mediaType != target.MediaType {
+		return nil, fmt.Errorf("%s %q: mismatch response Content-Type %q: expect %q", resp.Request.Method, resp.Request.URL, mediaType, target.MediaType)
+	}
+	if size := resp.ContentLength; size != -1 && size != target.Size {
+		return nil, fmt.Errorf("%s %q: mismatch Content-Length", resp.Request.Method, resp.Request.URL)
+	}
+	if err := verifyContentDigest(resp, target.Digest); err != nil {
+		return nil, err
+	}
+	return resp.Body, nil
+}
+
+// Push pushes the content, matching the expected descriptor.
+func (s *manifestStore) Push(ctx context.Context, expected ocispec.Descriptor, content io.Reader) error {
+	return s.repo.push(ctx, expected, content, expected.Digest.String())
+}
+
+// Exists returns true if the described content exists.
+func (s *manifestStore) Exists(ctx context.Context, target ocispec.Descriptor) (bool, error) {
+	_, err := s.Resolve(ctx, target.Digest.String())
+	if err == nil {
+		return true, nil
+	}
+	if errors.Is(err, errdef.ErrNotFound) {
+		return false, nil
+	}
+	return false, err
+}
+
+// Delete removes the content identified by the descriptor.
+func (s *manifestStore) Delete(ctx context.Context, target ocispec.Descriptor) error {
+	return s.repo.delete(ctx, "manifests", target)
+}
+
+// Resolve resolves a reference to a descriptor.
+func (s *manifestStore) Resolve(ctx context.Context, reference string) (ocispec.Descriptor, error) {
+	ref, err := s.repo.parseReference(reference)
+	if err != nil {
+		return ocispec.Descriptor{}, err
+	}
+	url := fmt.Sprintf("%s/manifests/%s", s.repo.endpoint(), ref.Reference)
+	req, err := http.NewRequestWithContext(ctx, http.MethodHead, url, nil)
+	if err != nil {
+		return ocispec.Descriptor{}, err
+	}
+	req.Header.Set("Accept", manifestAcceptHeader)
```

Comment:
Can you open an issue that would support exercising the registry down convertion path. For e.g. Request manifest even if the reference is actually a manifestList? 
Secondly also consider how we can pass in arbitary mediaTypes?

Reply:
Sure. I've opend #83 for your first question.
For your second question, I don't see a scenario for passing arbitary mediaTypes. Could you give one so that I can refine my design and implementation? The current design is that we should send a PR to update `manifestAcceptHeader` when there is a new manifest spec.

Reply:
The manifest mediaType could not be used. For e.g crane had this hardcoded and we couldn’t get the descriptor on resolve for an ORAS manifest . I would recommend supporting an option to provide custom accept headers.

Reply:
OK. Let's make it an option.

Reply:
Updated with 747c1c3